File: JS/Medium/BinaryTree/LevelOrderTraversal/rightSideView.js
/** 199
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
  if (!root) return [];
  let result = [];

  let q = [root];

  while (q.length) {
    let levelSize = q.length;
    for (let i = 0; i < levelSize; i++) {
      let curr = q.shift();
      if (i == levelSize - 1) result.push(curr.val);

      curr.left && q.push(curr.left);
      curr.right && q.push(curr.right);
    }
  }
  return result;
};

var rightSideView = function (root) {
  let res = [];

  let traverse = (curr, depth) => {
    if (!curr) return;
    if (res.length == depth) {
      res.push(curr.val);
    }

    traverse(curr.right, depth + 1);
    traverse(curr.left, depth + 1);
  };
  traverse(root, 0);
  return res;
};
let vals = [1, 2, 3, null, 5, null, 4];
console.log(rightSideView(vals));

File: C#/DSA/Medium/Trees/rightSideView.cs

// // Definition for a binary tree node.
// public class TreeNode
// {
//     public int val;
//     public TreeNode left;
//     public TreeNode right;
//     public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
//     {
//         this.val = val;
//         this.left = left;
//         this.right = right;
//     }
// }

// public class Solution
// {
//     public IList<int> RightSideView(TreeNode root)
//     {
//         IList<int> res = [];

//         Queue<TreeNode> q = new();
//         q.Enqueue(root);

//         while (q.Count > 0)
//         {
//             int levelSize = q.Count;
//             for (int i = 0; i < levelSize; i++)
//             {
//                 TreeNode curr = q.Dequeue();
//                 if (i == levelSize - 1) res.Add(curr.val);

//                 if (curr.left != null) q.Enqueue(curr.left);
//                 if (curr.right != null) q.Enqueue(curr.right);
//             }
//         }

//         return res;
//     }
// }